pytest_plugins/testimony_markers.py

```diff
@@ -0,0 +1,106 @@
+import inspect
+import logging
+import re
+
+import pytest
+
+LOGGER = logging.getLogger('robottelo')
+
+IMPORTANCE_LEVELS = []
+
+
+def pytest_addoption(parser):
+    """Add CLI options related to Testimony token based mark collection"""
+    parser.addoption(
+        '--importance',
+        help='Comma separated list of importance levels to include in collection',
+    )
+    parser.addoption(
+        '--component',
+        help="Comma separated list of component names to include in collection",
+    )
+
+
+def pytest_configure(config):
+    """Register markers related to testimony tokens"""
+    for marker in [
+        'importance: CaseImportance testimony token, use --importance to filter',
+        'component: Component testimony token, use --component to filter',
+        # TODO: components read from testimony.yaml
+    ]:
+        config.addinivalue_line("markers", marker)
+
+
+component_regex = re.compile(
+    # To match :CaseComponent: FooBar
+    r"\s*:CaseComponent:\s*(?P<component>\S*)",
+    re.IGNORECASE,
+)
+
+importance_regex = re.compile(
+    # To match :CaseImportance: Critical
+    r"\s*:CaseImportance:\s*(?P<importance>\S*)",
+    re.IGNORECASE,
+)
+
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_collection_modifyitems(session, items, config):
+    """Add markers for testimony tokens"""
+    # split the option string and handle no option, single option, multiple
+    # config.getoption(default) doesn't work like you think it does, hence or ''
+    importance = [i for i in (config.getoption('importance') or '').split(',') if i != '']
+    component = [c for c in (config.getoption('component') or '').split(',') if c != '']
+
+    selected = []
+    deselected = []
+    for item in items:
+        if item.nodeid.startswith('tests/robottelo/'):
+            # Unit test, no testimony markers
+            continue
+
+        # apply the marks for component and importance
+        # Find matches from docstrings starting at smallest scope
+        item_docstrings = [
+            d
+            for d in map(inspect.getdoc, (item.function, getattr(item, 'cls', None), item.module))
+            if d is not None
+        ]
+        for docstring in item_docstrings:
+            item_mark_names = [m.name for m in item.iter_markers()]
```

Comment:
I think this can be fetched above and outside of this for loop. Since the marks for item are not going to change for an item docstring. Saving the time a bit.